spatialos/UnrealGDK
# Locking gas ability activations

Tilman and Ally joint-effort.

[Matching 4.23 engine PR](https://github.com/improbableio/UnrealEngine/pull/333)

Below is a gif from a test gym where pressing a key starts a gameplay ability for 5 seconds where the Actor is locked during activation and unlocked when the activation ends:

![asdfasdfasdf](https://user-images.githubusercontent.com/9222722/73268668-4b923780-41d3-11ea-9455-a471ae3428b1.gif)

#### Description
This PR adds server-side locking for the duration of GAS GameplayAbility activation such that AuthorityIntent will not be changed while the server has an active GameplayAbility (this excludes inflight-but-not-yet-received ClientPredicted or LocalOnly abilities).

This is implemented predominantly in engine fork code also with the locking policy, which exposes an ability to acquire and release locks using a new engine delegate. This is done so that we can encapsulate the functionality within the engine AbilitySystemComponent and not need to introduce our own within the GDK.   

#### Remaining issues

The test gym usage is very trivial and other usages of gameplay abilities can break things. For example, the [GASDocumentation](https://github.com/tranek/GASDocumentation) test gym has a gameplay ability on the player state (pretty standard thing) but with the current implementation, when a player crosses a boundary while activating an ability, the player state is correctly locked but linked actors such as the controller and character migrate - this leads to bad behaviour and is the problem that Actor sets solve.

This is visible in the top-left authority icon colours in this gif:

![MeteorLocking](https://user-images.githubusercontent.com/9222722/73269102-1a663700-41d4-11ea-896a-2abd2f08b367.gif)

#### Tests
The exposed delegate functionality in the locking policy has an additional unit test for the reference-counted locking policy.

#### How can this be verified by QA?
Run the locking policy unit tests.
Open the UnrealGDKTestGym map (PR to be linked) and follow the instructions to mimic what you can see in the gif above.

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.
